### PR TITLE
Sprint5 fixes, inplementing is-uuid library

### PR DESCRIPTION
**Pull request by ZakBrinlee**
Relating to [Issue #54](https://github.com/ActoKids/api/issues/54)

**Testing**: 
Get all events: [https://api.2edusite.com/v1/events](https://api.2edusite.com/v1/events)
Get event by event_id: [https://api.2edusite.com/v1/events/{event_id}](https://api.2edusite.com/v1/events/94fd882e-4680-47fd-b107-320d86b6f7b8)

**Summary of changes**:
- Included necessary files for adding the is-uuid library to GET events by id function. 
- Changed code to work with callbacks instead of the proxy integration for error handling
- Changed table and region to Oregon/Production region

### CODE REVIEW

File: lambda/events/get/index.js [hunk 1]
```diff
@@ -1,78 +1,84 @@
...
-        TableName: 'ak-api-zak-dynamodb',
-        ProjectionExpression: "event_id, user_name, activity_type, description, org_name, location_name, location_address, contact_name, contact_phone, contact_email, start_date_time, end_date_time, frequency, cost, picture_url, min_age, max_age, disability_types, event_status, approver, created_timestamp, event_link, event_name"
-    }//end of params
+exports.handler = async (event, context, callback) => {
+  
+   // Console log to CloudWatch for Lambda function start
+   console.log("ActoKids GET all events function started.");
+  
+   // Instantiate new instance of DynamoDB object with specified api version.
+   const db = new AWS.DynamoDB({ apiVersion: '2012-10-08'});
```
Comment (MrSamsa42) [inline lambda/events/get/index.js]: db variable is assigned a value but never used.  This line and associated comment can be deleted.

File: lambda/events/get/index.js [hunk 2]
```diff
@@ -1,78 +1,84 @@
...
+       responseBody = data;
+       statusCode = 200;
+      
+       // Log for successful DynamoDB scan
+       console.log("Successful scan. Status code: " + statusCode)
 
+       callback(null, responseBody);
+   } catch(err) {
+       //setting responseBody and statusCode on successful return from DynamoDB
+       responseBody = 'Unable to connect to database';
```
Comment (MrSamsa42) [inline lambda/events/get/index.js:55]: API gateway is configured to look for the keyword, 'ResourceNotFoundException', in order to return the appropriate response (500).  Therefore, this variable is not needed and line 66 should be callback(JSON.stringify(err), null).

File: lambda/events/get/index.js [hunk 3]
```diff
@@ -1,78 +1,84 @@
...
+       statusCode = 500;
+      
+       // Log for error with DynamoDB scan
+       console.error("Unsuccessful connection to DynamoDB. Status code: " + statusCode);
+       console.error("Error: " + err);
+  
+       callback(JSON.stringify(responseBody), null)
+   }//end of catch
+
+   // Create response to return with statusCode, headers, and responseBody.
```
Comment (MrSamsa42) [inline lambda/events/get/index.js]: I'm pretty sure this code isn't doing anything, apart from being logged to Cloudwatch.  Headers are being set by API Gateway, so it's probably save to remove lines 70-78 and update Cloudwatch logs accordingly.

File: lambda/events_id/get/index.js [hunk 4]
```diff
@@ -2,104 +2,127 @@
...
 // event is the header and parameters from the request
-exports.handler = async (event, context) => {
+exports.handler = async (event, context, callback) => {
     
     // Console log to CloudWatch for Lambda function start
-    console.log("ak-api-zak-lambda GET event by event_id function started.");
+    console.log("ActoKids GET event by event_id function started.");
     
     // Insantiate new instance of DynamoDB object with specified api version. 
     const db = new AWS.DynamoDB({ apiVersion: '2012-10-08'});
```
Comment (MrSamsa42) [inline lambda/events_id/get/index.js]: db is assigned a value but never used.  This line and associated comment can be deleted.

File: lambda/events_id/get/index.js [hunk 5]
```diff
@@ -2,104 +2,127 @@
...
-        console.error("Error: " + err);
-    }//end of catch
+    //Log calling API type and endpoint
+    console.log("API-Endpoint called: " + httpMethod + " " + event.context['resource-path']);
+    
+    //Log calling IP
+    console.log("Calling-IP: " + event.params.header['X-Forwarded-For']);
 
     // Create response to return with statusCode, headers, and responseBody. 
     const response = {
```
Comment (MrSamsa42) [inline lambda/events_id/get/index.js:49]: I'm pretty sure this code isn't doing anything, apart from being logged to Cloudwatch. Headers are being set by API Gateway, so it's probably save to remove lines 52-57 and update Cloudwatch logs accordingly

File: lambda/events_id/get/index.js [hunk 6]
```diff
@@ -2,104 +2,127 @@
...
         },
-        body: responseBody
+        body: body
     };
+    
+    // validate input parameter as uuid version 5, when fails a 400 response must be given
+    if(!isUUID.anyNonNil(eventId)) {
+    	console.log("ERROR: uuid in not uuid/v4");
+    	response.statusCode = 400;
+		callback(JSON.stringify(response), null);
```
Comment (MrSamsa42) [inline lambda/events_id/get/index.js:58]: This is fine, but API Gateway needs to be configured to handle the response accordingly, based on the string 'statusCode = 400'

File: lambda/events_id/get/index.js [hunk 7]
```diff
@@ -2,104 +2,127 @@
...
+            // Creating variable {data} to store the response from DynamoDB using the params created above
+            // documentClient.query() will search the entire table based on the parameters given.
+            const data = await documentClient.query(params).promise();
+            response.body = data;
+            console.log(JSON.stringify(response));
+            // Checking if the query returned with an event or found no matches
+            if(response.body.Count == 0) {
+                //setting responseBody and statusCode on no results returned from DynamoDB
+                response.Body = 'Event not found';
+                response.statusCode = 404;
```
Comment (MrSamsa42) [inline lambda/events_id/get/index.js:87]: This is fine, but API Gateway needs to be configured to handle the response accordingly, based on the string 'statusCode = 404'

File: lambda/events_id/get/index.js [hunk 8]
```diff
@@ -2,104 +2,127 @@
...
+            }
+            
+        } catch(err) {
+            response.body = 'Failure to connect or retrieve from the database';
+            response.statusCode = 500;
+            
+            // Log for error with DynamoDB query
+            console.error("Unsuccessful connection to DynamoDB. Status code: " + statusCode);
+            console.error("Error: " + err);
+            callback(JSON.stringify(response), null)
```
Comment (MrSamsa42) [inline lambda/events_id/get/index.js:109]: This is fine, but API Gateway needs to be configured to handle the response accordingly, based on the string 'statusCode = 500'

File: lambda/events/get/index.js (hunk 2 above)
Comment (ZakBrinlee) [reply lambda/events/get/index.js:55]: I keep the variables for writing the full easy to read response for CloudWatch but making the changes for callback including the error.

File: lambda/events_id/get/index.js (hunk 4 above)
Comment (ZakBrinlee) [reply lambda/events_id/get/index.js]: fixed

File: lambda/events/get/index.js (hunk 1 above)
Comment (ZakBrinlee) [reply lambda/events/get/index.js]: fixed

File: lambda/events/get/index.js (hunk 3 above)
Comment (ZakBrinlee) [reply lambda/events/get/index.js]: Removed headers, kept the response to include in the Cloudwatch logs

File: lambda/events_id/get/index.js (hunk 5 above)
Comment (ZakBrinlee) [reply lambda/events_id/get/index.js:49]: Removed the code about header in the response. Keeping response as it is being used throughout the code and for my logging.

File: lambda/events/get/index.js [hunk 9]
```diff
@@ -1,78 +1,78 @@
...
-    
-    // Console log to CloudWatch for Lambda response
-    console.log("Response being returned by Lambda: ", response);
-    
-    // Console log to CloudWatch for Lambda function ending
-    console.log("ak-api-zak-lambda function ended.");
-    
-    return response;
-}//end of exports.handler
+   // Log calling API type and endpoint
```
Comment (Enshaedn) [inline lambda/events/get/index.js]: Technically this comment should be different than the comment on line 27. Should say something more to the effect of //Log calling IP.

File: lambda/events/get/index.js [hunk 10]
```diff
@@ -1,78 +1,78 @@
...
+   console.log("Calling from IP: " + event.params.header['X-Forwarded-For']);
+  
+   // parameters for DynamoDB request.
+   // ProjectionExpression is for the request attributes of events in table
+   const params = {
+       TableName: 'ak-prod-events-dynamo',
+       ProjectionExpression: "event_id, user_name, activity_type, description, org_name, location_name, location_address, contact_name, contact_phone, contact_email, start_date_time, end_date_time, frequency, cost, picture_url, min_age, max_age, disability_types, event_status, approver, created_timestamp, event_link, event_name"
```
Comment (Enshaedn) [inline lambda/events/get/index.js]: Need to add a parameter to this: "inclusive_event"

File: lambda/events_id/get/index.js [hunk 11]
```diff
@@ -2,104 +2,121 @@
...
+            // Specify which attribute to search for
+            ExpressionAttributeNames:{
+                "#event_id": "event_id"
+            },
+            // Specify which attribute value {event id} to search for
+            ExpressionAttributeValues: {
+                ":event_id": eventId
+            },
+            // ProjectionExpression is for the request attributes of events in table 
+            ProjectionExpression: "event_id, user_name, activity_type, description, organization_name, location_name, location_address, contact_name, contact_phone, contact_email, start_date_time, end_date_time, frequency, cost, picture_url, min_age, max_age, disability_types, show_status, approver"
```
Comment (Enshaedn) [inline lambda/events_id/get/index.js]: The projection expression appears to have some incorrect parameters.
- organization_name should be org_name
- show_status should be event_status
- missing inclusive_event

Comment (Enshaedn) [reply lambda/events_id/get/index.js]: Also missing event_link

**Review by Enshaedn — CHANGES_REQUESTED** [review]
Just need a few tweaks, but otherwise looks good.

File: lambda/events/get/index.js (hunk 9 above)
Comment (ZakBrinlee) [reply lambda/events/get/index.js]: fixed

File: lambda/events/get/index.js (hunk 10 above)
Comment (ZakBrinlee) [reply lambda/events/get/index.js]: Removed ProjectionExpression field so the QUERY will return all fields

File: lambda/events_id/get/index.js (hunk 11 above)
Comment (ZakBrinlee) [reply lambda/events_id/get/index.js]: Removed ProjectionExpression field so the QUERY will return all fields

**Review by Enshaedn — APPROVED** [review]
I approve of the changes. Removing the projection expression entirely will allow the GET endpoints to return all fields from Dynamo and let the UI determine what to display.

**Review by MrSamsa42 — APPROVED** [review]
It looks like all of my previous concerns have been addressed.

**Pull request merged**